File: hw/maisincho/u3/HW23Observer/Observer/src/main/java/ec/edu/espe/observer/model/Stock.java
package ec.edu.espe.observer.model;

import java.util.ArrayList;
import java.util.Iterator;

/**
 *
 * @author Isaac Maisincho Crafters_Market DCCO ESPE
 */
public abstract class Stock {

    private String symbol;
    private double price;
    private ArrayList<IInvestor> investors = new ArrayList<>();

    public Stock() {

    }

    public void addObserver(IInvestor iinvestor) {
        investors.add(iinvestor);
    }

    public void deleteObserver(IInvestor iinvestor) {
        investors.remove(iinvestor);
    }

    public void notifyObservers(Object args) {
        Iterator<IInvestor> i = investors.iterator();
        while (i.hasNext()) {
            IInvestor investor = i.next();
            investor.update(this, args);
        }
    }

    public abstract String getSymbol();

}

File: ws/jacome/U3/WS35Observer/JS/model/Stock.js
class Stock {
    constructor() {
        this.investors = [];
    }

    addObserver(investor) {
        this.investors.push(investor);
    }

    deleteObserver(investor) {
        const index = this.investors.indexOf(investor);
        if (index !== -1) {
            this.investors.splice(index, 1);
        }
    }

    notifyObservers(args) {
        for (const investor of this.investors) {
            investor.update(this, args);
        }
    }

    getSymbol() {
        throw new Error("Method 'getSymbol()' must be implemented.");
    }

    getPrice() {
        throw new Error("Method 'getPrice()' must be implemented.");
    }
}

module.exports = Stock;